### ISSUE TITLE
Routing: Display of route length is not really usable (when caching on a bike)

### ISSUE THREAD

**Issue by MagpieFourtyTwo** [cgeo/cgeo#15706]
### Describe your problem!

When caching on a bicycle, I really appreciate the possibility to see the routing distance to the target in big letters.
But there are some caveats:

1. while approaching the target, at some distance (which most of the times seems to be s/th around 200, but sometimes was even 900 meters!) the display starts to flicker, i. e. goes off/on/off - most probably to draw your attention, but although this is a pretty neat idea, the frequency of about a second is way too (s)low, because every time I am drawn to it, there is ... nothing, cause it's currently off. And esp. while riding in the woods one can't afford a whole second of inattention, just to wait for the numbers ...
2. even more as one can not even be sure that the numbers will come back at all, cause as soon as you get too close to the target, the route length will just go away ,,, and never come back again. Which in fact means even at least two seconds of inattention, till you recognize, which - while traveling at 20 km/h - means riding 5(!) bike lengths without even noticing obstacles ...

### How to reproduce?

1. route to any target 
2. touch display of route length twice to see the numbers in big
3. approach the target ...

### Actual result after these steps?

4. at some point route length display flickers (with way too low frequency)
5. at some further point display of route length disappears completely

### Expected result after these steps?

- either frequency should be way higher or have a different rhythm (e. g. 1 second on, 1/10 second off, 1 sec on ...)
- even better: route length should be readable all the time (and perhaps may just change font colors)
- just in case there is no route length (because of "no route found" or "too close to get a route"), straight distance should be shown
- (straight) distance should be shown even if it's zero, cause zero is not nothing, but an important information, which significantly differs from "not showing anything because there is no route".

### Reproducible

Yes

### c:geo Version

2024.05.07-NB-e2cd809

**Comment by moving-bits** [comment]
Strange, have never seen this yet. But anyway - FYI: There's no "flickering mode" built-in, so there has to be a different reason that causes routed distance to go away.

Which map type is affected (or all)?

**Comment by MagpieFourtyTwo** [comment]
Ok - I have never ever seen it in other than described, thus I thought this the way it was intended. Interesting. ;)

As for the maps: I use offline OSM most of the time, but can try with others, if of any interest. But I guess you meant UM vs. Classic? TBH, haven't put any effort in distinguishing, cause I thought this is normal anyway, but would say, they behave exactly the same. But will cross-check tomorrow.

**Comment by moving-bits** [comment]
Thanks. Yes, I meant UM vs. classic. Actually we do have three different map implementations currently - "classic Google maps" (aka "CGeoMap", the oldest map implementation), "classic OSM" (aka "NewMap") and "UnifiedMap" (which tries to use the same business code for both Google Maps and OSM wherever possible), though even here you have subtle differences due to different API and different behavior of the underlying libraries.

And even though we try to reuse some functions for all of these three map types (eg: map quick settings), you may still get some slightly different behavior, as some other related facts may differ (eg: when and where to put a location update notification into the maps' internal processings).

**Comment by MagpieFourtyTwo** [comment]
Both do the blinking.

And esp. when approaching the target, lets say the last about 100 m, big bike route length numbers usually disappear completely. But admittedly even the small ones are not visible anymore, thus most probably no route is calculated anymore at all.

Typical scenario:
- Target far away: everything fine, routing distance field (i. e. biker and big numbers) are constantly visible (as expected)
- at about 200 to 300 m left: field starts to flicker with a rhythm of about a second off, second on, second off ...
- about 50 to 100 m left: field disappears completely

**Comment by ztNFny** [comment]
Maybe instead of removing the view when distance can't be calculated it should just be out in a disabled state - or just wait before any changes for x seconds whether a new calculation result becomes available. 
Shouldn't matter too much if an outdated distance is shown for 2-3 seconds - gps always lags a bit anyway.

**Comment by MagpieFourtyTwo** [comment]
> Shouldn't matter too much if an outdated distance is shown for 2-3 seconds

Yes, but only if the distance to target (or to the next turn resp. waypoint) is greater than the one which will be passed within these 2-3 seconds (even on bike up to 25 meters, in car way more).  Would not be too good to miss a turn or, even worse, to take the next one by mistake, just because c:geo pretended at the previous one that there were still 200 meters to go ;)

Therefore I would suggest to show such "cached" distances e. g. in a different color (or alike), so the user can judge the (un)reliability of the shown value.

However, although already better, this will not solve the issue that the route resp. distance to target gets completely lost in the very last seconds, on approaching the target (so actually at the very moment when it is most important).  While it would even be worse if esp. for these last few meters the distance freezes.  So, in these "close to target" cases, I would prefer if the value shown will not ne the frozen one, but the one of the straight distance.

**Comment by MagpieFourtyTwo** [comment]
TMALSS: When route "gets lost", don't disable display, but with ...
... long distance to next point or target:  show "frozen" distance in different color
.. short distance to next point or target: show value (and perhaps even symbol) of straight line distance

**Comment by ztNFny** [comment]
> this will not solve the issue that the route resp. distance to target gets completely lost in the very last seconds, on approaching the target

Do you have any reliable steps to reproduce that? I'm using routing all the time, both walking and bike, and have never seen that.

A "cached" identifier would again to lead to flickering this issue is about, just a different style.

**Comment by MagpieFourtyTwo** [comment]
> Do you have any reliable steps to reproduce that?

Yep, tons of - at least with our devices over here (Huawei P30 Pro).  I just have to got out biking.  ;)
It happens every single time, and so consistently that I thought this bug, which it apparently seems to be, is a feature.
But perhaps it's just because of the "older" hardware ...?  Or perhaps due to bad GSM or GPS signal?
As we're usually riding in the woods ... hmm.  Will have a look at signal strengths, too.

Ok, scenario is always the same, as already described in the OP:
1. open listing of any cache and route to it (or one of it's stages) via compass symbol in header row
2. set routing style to "Bile" and click on the distance value fields to get the big numbers
3. mount your phone to the bike and go there with about 20 to 25 km/h
4. starting with a distance of a few hundred meters the flickering starts
5. somewhere within a distance of 100 meter or less the display vanishes completely

> I'm using routing all the time, both walking and bike, and have never seen that.

Strange.  But in fact the same as @moving-bits said.
Will add. cross-check with my new phone (Honor Magic6 Pro) - it's way faster on routing (who would have thought ;) ), so the flickering will perhaps not be visible anymore at least ...?  ;)

> A "cached" identifier would again to lead to flickering this issue is about, just a different style.

Yes, this would of course not work if it would be the only change, I had assumed (and have probably misunderstood your proposal) that you wanted to just not hide the routing distance, because of "no route", but leave it there - and as this most probably needs an explicit draw command (to replace the vanished number display with the numbers of the last calculation), I would then prefer the "cached" numbers in a different color.  But now it seems to me as if you just wanted to delay the next update ...?

**Comment by ztNFny** [comment]
> I had assumed (and have probably misunderstood your proposal) that you wanted to just not hide the routing distance, because of "no route", but leave it there - and as this most probably needs an explicit draw command (to replace the vanished number display with the numbers of the last calculation), I would then prefer the "cached" numbers in a different color. But now it seems to me as if you just wanted to delay the next update ...?

My understanding of your problem (simplified): Route distance can be calculated for 1 second, then for 1 second it can't, next second it's available again, ... leading to 1 Hz flickering because the distance indicator just disappears as soon as no routing can be calculated.
Now if there was a small grace period before the indicator disappeared - in the above example 1s would be enough - you wouldn't have any flickering info anymore.

But really, we should figure out why flickering is so bad for you, then we probably don't need to change anything about flickering.

**Comment by MagpieFourtyTwo** [comment]
> Route distance can be calculated for 1 second, then for 1 second it can't, next second it's available again, ... leading to 1 Hz flickering because the distance indicator just disappears as soon as no routing can be calculated.

Exactly.

Yep, consented.  However, a grace period (read: lower recalculation frequency) might still be a good idea.  This could perhaps even prevent (or at least reduce the likelihood of) this "route jumping" that sometimes occurs, i. e. the route changes direction by 90 degrees just because you pass under a bridge.

**Comment by ztNFny** [comment]
> . This could perhaps even prevent (or at least reduce the likelihood of) this "route jumping" that sometimes occurs, i. e. the route changes direction by 90 degrees just because you pass under a bridge.

That'd require somehow remembering the previous route and comparing it with the new one, only accepting new if it's similar enough. Don't think that's really feasible - and if it is it should be part if Brouter and not cgeo.

Anyway, let's try to get a reliably - and ideally in emulator - reproducable scenario for your "route doesn't calculate" problem.

**Comment by MagpieFourtyTwo** [comment]
> That'd require somehow remembering the previous route and comparing it with the new one, only accepting new if it's similar enough.

Interesting approach - and way more elaborated than what I had in mind:
I simply thought of just decreasing the frequency of route recalculation.  This way it will perhaps gain some "stability" resp. is not that "jumpy" like it's currently from time to time.

**Comment by MagpieFourtyTwo** [comment]
Re-evaluated the "blinking route distance on last few hundred meters" - and it even happens with the new hardware, thus it's most definitely not a problem of CPU capacity.

**Comment by MagpieFourtyTwo** [comment]
It's still happening - and today I made another observation that also seems to prove that the "vanishing" resp. blinking route length display is NOT a performance issue: By accident, I stopped in the very moment that the route length display vanished. (Add. remark: In the pic below the bike route length should be displayed in big numbers).

But although I stayed there for a minute or so, the route length display did not come back. Moreover: Although no length is shown (i. e. seemingly no route has been calculated), on the map there *is* a "correct" route drawn (i. e. it's *not* "invisible" and not a straight line, instead it follows the road curve as it should):

![No route length display - although route is visible](https://github.com/user-attachments/assets/68c85f3c-f59c-4b5e-b564-44baa03572ec)

This BTW does not only happen with routing strategy "bike" but with "car", too - you only have to drive slow enough to notice.  ;)

**Comment by moving-bits** [comment]
are you using "show both distances" setting? If so, can you test with deactivated setting?

**Comment by MagpieFourtyTwo** [comment]
Yep, I have ``Show straight distance`` activated.
Will check how it behaves with straight distance deactivated.

**Comment by moving-bits** [comment]
Note to myself: possible cause (to be verified in code):
- routed and straight distance may give the same value when close to target
- then only one of the two will be displayed
- if that's the one currently being supersized, supersize will be removed

**Comment by MagpieFourtyTwo** [comment]
Chapeau - good idea, sounds pretty reasonable.
BTW: If the above is really the case, ``Straight distance`` seems to always be the winner.
That's why I am using them for the big numbers (once the blinking has annoyed me too much again),
cause they have never left me alone so far. ;)

**Comment by MagpieFourtyTwo** [comment]
Ok, tested today with ``Show straight distance`` disabled (but still with an individual route set and visible):
On approach of the target the big number display did not "blink" anymore,
but changed back and forth between straight distance and routing distance.
Which is in fact just the way I would have expected (and even proposed, IIRC) it to be.  ;)

So with ``Show straight distance`` activated, the only difference should be that the straight distance should *additionally* be shown in "small numbers" in the upper right corner.

There's just one case where things can become a bit ... hectic. Scenario:
- routing is active, route distance is displayed as big number
- when no route could be calculated, the straight distance will be displayed instead of the routing distance
- foll. the usual strategy, the straight distance will no longer be displayed in small in the upper right corner

... but I would suggest to just leave it there as it is, at least in this case, cause most probably in the very next second a route can be calculated again - and then the big number would switch to routing distance, and straight distance must be displayed in small again. Which, esp. together with an individual route, would create some hectic on the display, with jumping numbers for at eleast the distance of the individual route ... that's why I would just leave the straight distance as it is, i. e. *always* show it, if routing is active.

**Comment by moving-bits** [comment]
Proposal:

- Rearrange the distance views for easier window handling
- Put the (up to) three distance views below target & supersize view
- Always display both straight + routed distance (if enabled), even if one of them is shown in supersize view

Could look like this (crude mockup):
![image](https://github.com/user-attachments/assets/564ceaef-dada-49dc-8799-9a97c1e58a70)

- target info would get a full line (visible only if target set)
- overlaps with supersize view (as it does today); supersize can be off, straight or routed distance (as today)
- individual route distance is left-aligned below target info / supersize view (visible only if individual route is active)
- routed distance is right-aligned below target info / supersize view (visible only of target is selected)
- straight distance is centered below target info / supersize view (visible only if target is selected + "show straight distance" is enabled

This would
- avoid the flickering described in this issue
- assign fixed horizontal positions to each distance view
- make window handling for distance views easier

Alternative idea:
Basically the same, but put the three distance view right-aligned below target info/supersize view.

**Comment by MagpieFourtyTwo** [comment]
Good ideas - and I would like the "Alternative idea" better, because I feel it easier to read when all the values are together:

![Routing Distances](https://github.com/user-attachments/assets/742a04e2-0c3e-4e21-9f5d-bcbd4dc2dc31)

But I guess getting used to the horizontal layout would just be a matter of a few days, so feel free to do what's easier to implement.  ;)

P.S.: If you prefer the horizontal display, you may want to keep the compass rose of map rotation in mind.

**Comment by ztNFny** [comment]
I once again managed to create a round trip, by moving 2 waypoints in the edit route dialog. but this time it was persistent, going back to list and map again didn't fix it this time, usually does. 
I then remembered a comment about "set start as target", so I disabled that, went to list view and back to map view and now the route was proper again.

**Comment by moving-bits** [comment]
> Good ideas - and I would like the "Alternative idea" better, because I feel it easier to read when all the values are together:

PR #16103 refactors distance views to be static TextViews below target / supersize info instead of dynamically creating/removing them. It implements the "alternative" version, meaning: up to three small distance views stacked vertically.

Besides code simplification that comes with that solution: Let's see if that also helps with the original issue of vanishing distance info in your environment.

**Comment by MagpieFourtyTwo** [comment]
Thx!  Although, the more I think about it, the more I ask myself, why - when beeing in routing mode - the routing distance should be overruled by the straight distance, just because they have the same value.  IMO, if I am in routing mode anyway, the most interesting value *is* the routing distance, and why should I care for that straight just gives the same value - by accident?

TMALSS, while beeing in routing mode, routing distance should have highest priority and should not be replaced or overruled by straight distance at all.

**Comment by MagpieFourtyTwo** [comment]
Any chance of getting this fixed soon?  Because disabling ``Show straight distance`` is not the real solution I was hoping for, cause without this setting, in ``Straight`` routing mode nothing is displayed at all (which can even be considered a bug, since in ``Straight`` the straight distance IMO should be treated as the calculated distance, so should be displayed regardless of this setting).

**Comment by moving-bits** [comment]
> in `Straight` routing mode nothing is displayed at all

PR #16156 fixes this part

**Comment by MagpieFourtyTwo** [comment]
And - following [Wilhelm Busch](http://www.wilhelm-busch-seiten.de/gedichte/schein11.html) ;) - how about the initial problem, i. e. preferring the routing distance over the straight distance if they're equal?  ;)

**Comment by moving-bits** [comment]
Problem is that I have never been able to reproduce this issue even for a single time, that makes it hard to fix. Code-wise it looks fine, and my hope was, that the last change (PR #16103) fixed flickering by no longer adding and removing views programmatically.

**Comment by MagpieFourtyTwo** [comment]
For at least me it's very "easy" to reproduce (as my typical using scenario seems to be very prone to this issue):

1. set ``[X] Show straight distance``
2. select a target as close as possible to a long, straight, drive-able (resp. route-able) path or street,
    where the last few 100 meters of the route are pretty straight (i. e. match the straight distance)
3. click the distance fields until the big numbers show the routing distance
4. drive on ... and once you've reached these last straight few 100 meters, keep an eye on the display:
   As soon as routing distance equals straight distance, the big numbers (with routing distance) will vanish.

For me this "works" reproducible and every single time.  :/

BTW, the reason why I like to additionally see the straight distance is, that routing sometimes goes weird ways ... and the comparison of routing and straight distance gives a good impression of the "route quality".

**Comment by moving-bits** [comment]
Thanks for the detailed description. I tried to reproduce it with this scenario, but failed so far :-(

c:geo always shows the distance for me, even if routed and straight distance seem to be the same (although both are floats, so chances are very small that they are actually the same. And, since last update, it does not matter anymore, at least not for supersize window, as the selected distance gets displayed even if routed and straight distance are the same).

But there's one BUT: If routed distance is less than a given minimum distance (which is set to 50 cm), then no value is shown. There's only one realistic scenario where this can happen, AFAICS: If there is no routed distance available for whatever reason, then routed distance is 0, which is less than allowed minimum.

Do you have some power saving feature activated for GPS signal?

**Comment by MagpieFourtyTwo** [comment]
> f there is no routed distance available for whatever reason, then routed distance is 0, which is less than allowed minimum.

Hmm, good point! My c:geo actually runs in energy-saving mode, which may have this effect, I guess. But I can turn it off for the sake of more GPS readings - it's not really necessary anyway (my battery lasts a good two days with normal use anyway), but I still like to be on the economical side. ;)  But if that's the price ... I'll test it out.  Thx so far!

**Comment by moving-bits** [comment]
Any new experiences with having GPS energy saving switched off?

And, BTW, which map types are affected - still all of old Mapsforge, old GM and UnifiedMap?

**Comment by MagpieFourtyTwo** [comment]
Not yet. Last weekend I could not reproduce, and this week there was no real chance to check, but I still have it in mind and will come back to this topic as soon as I have new insights, of course. :)

**Comment by MagpieFourtyTwo** [comment]
Status update: Blinking did not occur in the last months, so I would consider this solved.

Thx!

**Comment by MagpieFourtyTwo** [comment]
Addendum: Works even with energy saving mode on, used maps are usually OSM offline, most of the time "combined", UM/VTM, and sometimes GSat.
